permissions/Secrets.permission

```diff
@@ -0,0 +1,17 @@
+# -*- mode: sh -*-
+
+# x-sailjail-translation-catalog =
+# x-sailjail-translation-key-description =
+# x-sailjail-description = Secrets
+# x-sailjail-translation-key-long-description =
+# x-sailjail-long-description = Use DBus API to access the Sailfish Secrets daemon
+
+mkdir     ${RUNUSER}/sailfishsecretsd
+whitelist ${RUNUSER}/sailfishsecretsd
+read-only ${RUNUSER}/sailfishsecretsd
+
+# Allow discovery of the p2p address used by the daemon.
+dbus-user.talk      org.sailfishos.secrets.daemon.discovery
+dbus-user.call      org.sailfishos.secrets.daemon.discovery=org.sailfishos.secrets.daemon.*@/*
```

Comment:
Could we limit the methods or paths that are available or is this such interface that we can allow everything?

I know there are other places that we use full wildcards but in general we should pay more attention to these.

Reply:
The only method that is required is defined like that in the code:
```
    QDBusInterface iface("org.sailfishos.secrets.daemon.discovery",
                         "/Sailfish/Secrets/Discovery",
                         "org.sailfishos.secrets.daemon.discovery",
                         QDBusConnection::sessionBus());
    if (iface.isValid()) {
        QDBusReply<QString> reply = iface.call("peerToPeerAddress");
```
I'm not an expert in DBus archanes and not at all in firejail rules about DBus but after reading a bit the manual, I guess the two rules are fine:
```
dbus-user.talk      org.sailfishos.secrets.daemon.discovery
dbus-user.call      org.sailfishos.secrets.daemon.discovery=org.sailfishos.secrets.daemon.discovery@/Sailfish/Secrets/Discovery
```

I've updated the PR accordingly.

Reply:
[Firejail documentation](https://firejail.wordpress.com/features-3/man-firejail-profile/) is a bit terse on this but [xdg-dbus-proxy](https://manpages.debian.org/stretch-backports/xdg-dbus-proxy/xdg-dbus-proxy.1.en.html) explains these a bit better. Xdg-dbus-proxy is what firejail will ultimately use and although firejail also parses those rules they are passed to xdg-dbus-proxy in the same format. So for `talk` you need to use the service's bus name which is correct here, and then for `call` you need to have \<bus name\>=\<method name\>/\<path of the object\> that needs a little bit adjusting now. In this case `call` rule should be like this:
```
dbus-user.call      org.sailfishos.secrets.daemon.discovery=org.sailfishos.secrets.daemon.discovery.peerToPeerAddress@/Sailfish/Secrets/Discovery
```
Similarly for `broadcast` you can limit it to certain signals by specifying the signal names (using multiple rules if needed).

Reply:
Ah, thanks for the explanation. It makes a bit more sense for me now.

I'm still surprised that my previous rule, without the proper method name, was actually working. In fact, even removing the call rule completely still works…

Reply:
Well, `talk` rule without `call` rule would work because then you don't limit the methods you can make a call to.

I'm not sure how the version you have in the comment above could work. Are you sure it was sandboxed? I didn't try it myself though.

Reply:
Yes, here is my test case (slightly unrealistic, but easy to use) :
```
QT_LOGGING_RULES="org.sailfishos.*=true" /usr/bin/sailjail /usr/bin/gpg2 -s ~/Documents/toto
```
with:
```
[nemo@Sailfish devel]$ cat /usr/share/applications/gpg2.desktop 
[Desktop Entry]
Type=Application
Name=Mail
Icon=icon-launcher-email
Exec=/usr/bin/gpg2 -s /home/nemo/Documents/toto
X-MeeGo-Logical-Id=mail-ap-name
X-MeeGo-Translation-Catalog=jolla-email
X-Maemo-Service=com.jolla.email.ui
X-Maemo-Method=com.jolla.email.ui.activateWindow
X-Maemo-Object-Path=/com/jolla/email/ui

X-Desktop-File-Install-Version=0.26

[X-Sailjail]
Permissions=Accounts;Contacts;Phone;Email;WebView;Privileged;Internet;AppLaunch;Calendar;UserDirs;RemovableMedia;MediaIndexing;Sharing
OrganizationName=com.jolla
ApplicationName=email
```
And the PR moving the sailfish secret socket applied.

When working, the output starts with:
```
[D] unknown:0 - Connecting to secrets daemon via p2p address: "unix:path=/run/user/100000/sailfishsecretsd/p2pSocket" with connection name: "sailfishsecretsd-connection-0"
[D] unknown:0 - Connected to secrets daemon via connection: "sailfishsecretsd-connection-0"
```
When not working, it starts with :
```
[D] unknown:0 - Unable to query the peer to peer socket address from the secrets daemon!  Using fallback address.
```